consensus/src/creation/mod.rs

```diff
@@ -104,69 +139,95 @@ async fn create_unit<H: Hasher>(
 /// Section 5.1 for a discussion of this component.
 pub async fn run<H: Hasher>(
     conf: Config,
-    io: IO<H>,
+    mut io: IO<H>,
     mut starting_round: oneshot::Receiver<Option<Round>>,
     mut terminator: Terminator,
 ) {
+    futures::select! {
+        _= read_starting_round_and_run_creator(conf, &mut io, &mut starting_round).fuse() =>
+            debug!(target: "AlephBFT-creator", "Creator is about to finish."),
+        _ = terminator.get_exit() =>
+            debug!(target: "AlephBFT-creator", "Received an exit signal."),
+    }
+
+    terminator.terminate_sync().await;
+}
+
+async fn read_starting_round_and_run_creator<H: Hasher>(
+    conf: Config,
+    io: &mut IO<H>,
+    starting_round: &mut oneshot::Receiver<Option<Round>>,
+) {
+    let maybe_round = starting_round.await;
+    let starting_round = match maybe_round {
+        Ok(Some(round)) => round,
+        Ok(None) => {
+            warn!(target: "AlephBFT-creator", "None starting round provided. Exiting.");
+            return;
+        }
+        Err(e) => {
+            error!(target: "AlephBFT-creator", "Starting round not provided: {}", e);
+            return;
+        }
+    };
+
+    if let Err(err) = run_creator(conf, io, starting_round).await {
+        match err {
+            CreatorError::OutChannelClosed(e) => {
+                warn!(target: "AlephBFT-creator", "Notification send error: {}. Exiting.", e)
+            }
+            CreatorError::ParentsChannelClosed => {
+                debug!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.")
+            }
+        }
+    }
+}
+
+async fn run_creator<H: Hasher>(
+    conf: Config,
+    io: &mut IO<H>,
```

Comment:
This is for the purpose of the `terminator`. Channels we receive from outside should be kept till `terminate_sync` returns, otherwise other modules might start complaining about closed channels. There are some more details about that `terminator` here which are not clear - I will talk with its author.

Reply:
Oh, right, damn.